src/core/CORE_CONVENTIONS.md

```diff
@@ -0,0 +1,115 @@
+- [Core Conventions](#core-conventions)
+  - [1. Exposing API Types](#1-exposing-api-types)
+  - [2. API Structure and nesting](#2-api-structure-and-nesting)
+  - [3. Tests and mocks](#3-tests-and-mocks)
+
+# Core Conventions
+
+This document contains conventions for development inside `src/core`. Although
+many of these might be more widely applicable, adoption within the rest of
+Kibana is not the primary objective.
+
+## 1. Exposing API Types
+The following section applies to the types that describe the entire surface
+area of Core API's and does not apply to internal types.
+
+ - 1.1 All API types must be exported from the top-level `server` or `public`
+   directories.
+   
+   ```ts
+   // -- good --
+   import { IRouter } from 'src/core/server';
+
+   // -- bad --
+   import { IRouter } from 'src/core/server/http/router.ts';
+   ```
+   
+   > Why? This is required for generating documentation from our inline
+   > typescript doc comments, makes it easier for API consumers to find the
+   > relevant types and creates a clear distinction between external and
+   > internal types.
+ 
+ - 1.2 Classes must not be exposed directly. Instead, use a separate type,
+   prefixed with an 'I', to describe the public contract of the class.
+   
+   ```ts
+   // -- good (preferred) --
+   /**
+    * @public
+    * {@link UiSettingsClient}
+    */
+   export type IUiSettingsClient = PublicContractOf<UiSettingsClient>;
+
+   /** internal only */
+   export class UiSettingsClient {
+     constructor(private setting: string) {}
+     /** Retrieve all settings */
+     public getSettings(): { return this.settings; }
+   };
+
+   // -- good (alternative) --
+   export interface IUiSettingsClient {
```

Comment:
I'm not sure if there are cases where we prefer to have a dedicated interface instead of inferring a type from a class? If we don't have good examples of where this would be better, I can remove this from the conventions.

Reply:
Pros having explicit interfaces:
There is a difference in that an interface is a source of truth when we define it explicitly. it means that we have control of what properties are a part of a public contract.
Better tooling, as there are no intermediate types. Whenever I click `Go to definition` on `SavedObjectsClientContract` it navigates to `Pick` type.

Cons:
It creates an additional burden to define contracts manually.
Separates docs for an interface and implementation.

Reply:
Personally, I prefer explicit interfaces. I'd rather be deliberate about choosing to expose a property or method rather than allowing it to happen without notice (though the api-extractor tooling _may_ make it more obvious).

Reply:
I'm not completely convinced by the benefit of a separate interface. Adding a public method to an interface doesn't make it more obvious that it will change the API contract than adding a public method to a class.

It seems like typescript support for private named fields will land in February https://github.com/microsoft/TypeScript/wiki/Roadmap#38-february-2020. If classes support real private fields/methods it will remove the need for an intermediate type (whether an explicit interface or `PublicContractOf`). I see real private fields as the best solution, source code and docs live next to each other, no intermediate types, no type duplication. When Kibana adopts TS 3.8 would we still prefer a separate interface?

Reply:
> When Kibana adopts TS 3.8 would we still prefer a separate interface?

Don't think so. IIRC public/private field separation was the main reason for an explicit interface introduction.

Reply:
I think it could be slightly easier to migrate from `PublicContractOf` to private fields, but it wouldn't be that much work to migrate away from separate interfaces either.

Since we're hopefully adopting a new convention soon, it probably doesn't matter too much that we have two alternatives. I can flesh out the comment section to mention the pro's and cons of both conventions and mention that we're hoping to adopt private fields once it's available.

Reply:
Works for me. Agreed that in most use-cases real private fields works well.